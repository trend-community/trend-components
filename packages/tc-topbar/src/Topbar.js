import React, { Component, createRef } from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
import throttle from 'lodash/fp/throttle';

import { classNames } from './constants';

const TIMEOUT = 100;
const SCROLL_THRESHOLD = 24;

function defaultStyles() {
  return {
    backfaceVisibility: 'invisible',
    transform: 'translateY(0)',
    transition: 'transform 0.2s linear'
  };
}

function propGetter(defaultProps = {}) {
  return (props = {}) => ({
    ...props,
    className: cn(defaultProps.className, props.className)
  });
}

function getSectionProps(props = {}) {
  return {
    ...props,
    className: cn(classNames.SECTION, props.className, {
      [classNames.SECTION_START]: props.position === 'start',
      [classNames.SECTION_END]: props.position === 'end'
    })
  }
}

class Topbar extends Component {
  static propTypes = {
    children: PropTypes.func,
    className: PropTypes.string,
    compact: PropTypes.bool,
    fixed: PropTypes.bool,
    fixedScroll: PropTypes.bool,
    scrollTarget: PropTypes.object,
    tall: PropTypes.bool
  }

  static defaultProps = {
    scrollTarget: window
  }

  state = {
    height: 0,
    top: 0
  }

  constructor(props) {
    super(props);
    this.el = createRef();

    // Track some state internally to avoid unecessary renders.
    this.lastScrollTop = 0;
    this.currentScrollTop =

    this._scrollHandler = throttle(TIMEOUT)(this._scrollHandler.bind(this));
  }

  componentDidMount() {
    const { fixed, fixedScroll, scrollTarget } = this.props;

    const { scrollTop } = document.documentElement;

    this.lastScrollTop = scrollTop;
    this.currentScrollTop = scrollTop;

    this.setState({
      height: this.el.current && this.el.current.offsetHeight
    });

    scrollTarget.addEventListener('scroll', this._scrollHandler);
  }

  componentWillUnmount() {
    const { fixed, fixedScroll, scrollTarget } = this.props;

    scrollTarget.removeEventListener('scroll', this._scrollHandler);
  }

  api() {
    return {
      getElementProps: this._getElementProps,
      getInnerProps: propGetter({ className: classNames.INNER }),
      getSectionProps: getSectionProps,
      getIconProps: propGetter({ className: classNames.ICON }),
      getTitleProps: propGetter({ className: classNames.TITLE })
    }
  }

  render() {
    return this.props.children(this.api());
  }

  _fixedScrolling() {
    const { height } = this.state;
    const scrollPos = this.currentScrollTop;
    const adjustedPos = scrollPos <= height ? scrollPos * -1 : height * -1;
    const realPos = scrollPos < this.lastScrollTop ? 0 : adjustedPos;

    window.requestAnimationFrame(() => {
      this.el.current.style.transform = `translateY(${realPos}px)`;
      this.lastScrollTop = scrollPos;
    });
  }

  _getElementProps = (props = {}) => {
     const {
      compact,
      fixed,
      fixedScroll,
      tall,
      scrollTarget,
      ...rest } = this.props;
    const { isScrolling } = this.state;

    return {
      ...rest,
      ...props,
      className: cn(
        classNames.BASE,
        this.props.className,
        props.className,
        {
          [classNames.COMPACT]: compact,
          [classNames.FIXED]: fixed || fixedScroll,
          [classNames.FIXED_SCROLLING]: isScrolling && !fixedScroll,
          [classNames.TALL]: tall && !isScrolling
        }
      ),
      style: { ...defaultStyles(), ...this.props.style, ...props.style, },
      ref: this.el
    };
  }

  _scrollHandler = () => {
    const { fixed, fixedScroll } = this.props;

    if (!fixed && !fixedScroll) return;

    this.currentScrollTop = document.documentElement.scrollTop;

    const { height } = this.state;
    const isScrolling = this.currentScrollTop > SCROLL_THRESHOLD;
    const isStateScrolling = this.state.isScrolling;
    const isRealScrolling = isScrolling && !isStateScrolling;
    const isNoRealScrolling = !isScrolling && isStateScrolling;

    if (isRealScrolling || isNoRealScrolling) {
      this.setState({
        isScrolling: this.currentScrollTop > SCROLL_THRESHOLD
      });
    }

    fixedScroll && this._fixedScrolling();
  }
}

export default Topbar;
