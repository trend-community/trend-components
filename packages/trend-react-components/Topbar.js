export { default } from './src/Topbar';
