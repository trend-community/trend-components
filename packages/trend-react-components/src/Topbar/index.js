import React, { Component, createRef } from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
import throttle from 'lodash/fp/throttle';

import { classNames } from './constants';

const TIMEOUT = 100;
const SCROLL_THRESHOLD = 24;

function defaultStyles() {
  return {
    backfaceVisibility: 'invisible',
    transform: 'translateY(0)',
    transition: 'transform 0.2s linear'
  };
}

function propGetter(defaultProps = {}) {
  return (props = {}) => ({
    ...props,
    className: cn(defaultProps.className, props.className)
  });
}

function getSectionProps(props = {}) {
  return {
    ...props,
    className: cn(classNames.SECTION, props.className, {
      [classNames.SECTION_START]: props.position === 'start',
      [classNames.SECTION_END]: props.position === 'end'
    })
  }
}

class Topbar extends Component {
  static propTypes = {
    children: PropTypes.func,
    className: PropTypes.string,
    compact: PropTypes.bool,
    fixed: PropTypes.bool,
    fixedScroll: PropTypes.bool,
    scrollTarget: PropTypes.object,
    tall: PropTypes.bool
  }

  static defaultProps = {
    scrollTarget: window
  }

  state = {
    height: 0,
    lastScrollPos: 0,
    top: 0
  }

  constructor(props) {
    super(props);
    this.el = createRef();
    this._scrollHandler = throttle(TIMEOUT)(this._scrollHandler.bind(this));
  }

  componentDidMount() {
    const { fixed, fixedScroll, scrollTarget } = this.props;

    this.setState({
      height: this.el.current && this.el.current.offsetHeight,
      scrollPos: document.documentElement.scrollTop
    });

    if (fixed || fixedScroll) {
      scrollTarget.addEventListener('scroll', this._scrollHandler);
    }
  }

  componentWillUnmount() {
    const { fixed, fixedScroll, scrollTarget } = this.props;

    if (fixed || fixedScroll) {
      scrollTarget.removeEventListener('scroll', this._scrollHandler);
    }
  }

  _scrollHandler = () => {
    const { fixed, fixedScroll } = this.props;

    if (fixed || fixedScroll) {
      const scrollPos = document.documentElement.scrollTop;

      this.setState({
        scrollPos,
        isScrolling: scrollPos > SCROLL_THRESHOLD
      });
    }
  }

  componentDidUpdate(prevProps, prevState) {
    if (this.props.fixedScroll) {
      const prevScrollPos = prevState.scrollPos;
      const { height, scrollPos } = this.state;

      const pos = scrollPos < prevScrollPos
        ? 0
        : scrollPos <= height ? scrollPos * -1 : height * -1;

      window.requestAnimationFrame(() => {
        this.el.current.style.transform = `translateY(${pos}px)`;
      });
    }
  }

  _getElementProps = (props = {}) => {
     const {
      compact,
      fixed,
      fixedScroll,
      tall,
      scrollTarget,
      ...rest } = this.props;
    const { isScrolling } = this.state;

    return {
      ...rest,
      ...props,
      className: cn(
        classNames.BASE,
        this.props.className,
        props.className,
        {
          [classNames.COMPACT]: compact,
          [classNames.FIXED]: fixed || fixedScroll,
          [classNames.FIXED_SCROLLING]: isScrolling && !fixedScroll,
          [classNames.TALL]: tall && !isScrolling
        }
      ),
      style: { ...defaultStyles(), ...this.props.style, ...props.style, },
      ref: this.el
    };
  }

  getHelpers() {
    return {
      getElementProps: this._getElementProps,
      getInnerProps: propGetter({ className: classNames.INNER }),
      getSectionProps: getSectionProps,
      getIconProps: propGetter({ className: classNames.ICON }),
      getTitleProps: propGetter({ className: classNames.TITLE })
    }
  }

  render() {
    return this.props.children(this.getHelpers());
  }
}

export default Topbar;
