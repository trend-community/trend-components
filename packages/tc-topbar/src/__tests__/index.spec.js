import React from 'react';
import { mount, shallow } from 'enzyme';

import Topbar from '../Topbar';

describe('tc-topbar', () => {
  [
    'getElementProps',
    'getInnerProps',
    'getSectionProps',
    'getIconProps',
    'getTitleProps'
  ].forEach(propGetter => {
    it(`should expose ${propGetter}.`, () => {
      expect(setUp()[propGetter]()).toMatchSnapshot();
    });
  });

  it('should call the child function.', () => {
    const { childSpy } = setUp();
    expect(childSpy).toHaveBeenCalled();
  });

  it('should add scroll listener for "fixed" Topbar.', () => {
    const scrollTarget = { addEventListener: jest.fn() };
    const { wrapper } = setUp({ scrollTarget, fixed: true });
    expect(wrapper.prop('scrollTarget').addEventListener).toHaveBeenCalled();
  });

  it('should add scroll listener on scrollTarget by default.', () => {
    const scrollTarget = { addEventListener: jest.fn() };
    const { wrapper } = setUp({ scrollTarget, fixedScroll: true });
    expect(wrapper.prop('scrollTarget').addEventListener).toHaveBeenCalled();
  });
});

function setUp({ children = () => <div />,...props } = {}) {
  let renderArg;
  const childSpy = jest.fn(controllerArg => {
    renderArg = controllerArg;
    return children(controllerArg);
  });
  const wrapper = mount(<Topbar { ...props }>{ childSpy }</Topbar>);

  return { childSpy, wrapper, ...renderArg };
}
