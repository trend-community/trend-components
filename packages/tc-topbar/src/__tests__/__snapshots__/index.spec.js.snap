// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tc-topbar should expose getElementProps. 1`] = `
Object {
  "children": [MockFunction] {
    "calls": Array [
      Array [
        Object {
          "getElementProps": [Function],
          "getIconProps": [Function],
          "getInnerProps": [Function],
          "getSectionProps": [Function],
          "getTitleProps": [Function],
        },
      ],
      Array [
        Object {
          "getElementProps": [Function],
          "getIconProps": [Function],
          "getInnerProps": [Function],
          "getSectionProps": [Function],
          "getTitleProps": [Function],
        },
      ],
    ],
    "results": Array [
      Object {
        "isThrow": false,
        "value": <div />,
      },
      Object {
        "isThrow": false,
        "value": <div />,
      },
    ],
  },
  "className": "tc-Topbar",
  "ref": Object {
    "current": null,
  },
  "style": Object {
    "backfaceVisibility": "invisible",
    "transform": "translateY(0)",
    "transition": "transform 0.2s linear",
  },
}
`;

exports[`tc-topbar should expose getIconProps. 1`] = `
Object {
  "className": "tc-Topbar-icon",
}
`;

exports[`tc-topbar should expose getInnerProps. 1`] = `
Object {
  "className": "tc-Topbar-inner",
}
`;

exports[`tc-topbar should expose getSectionProps. 1`] = `
Object {
  "className": "tc-Topbar-section",
}
`;

exports[`tc-topbar should expose getTitleProps. 1`] = `
Object {
  "className": "tc-Topbar-title",
}
`;
